src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,261 @@
+        optionalTags.stream()
+                .filter(Optional::isPresent)
+                .forEach(tag -> giftCertificateDto.getTags().add(tag.get()));
+        return giftCertificateDto;
+    }
+
+    private void updateCertificateTags(Set<Tag> tags, long giftCertificateId) {
+        for (Tag tag : tags) {
+            Optional<Tag> tagOptional = tagDao.findByName(tag.getName());
+            Tag tagSureExist = tagOptional.orElseGet(() -> createCertificateTag(tag));
+            long tagId = tagSureExist.getId();
+            if (!giftCertificateHasTagDao.findTagIdsByCertificateId(giftCertificateId).contains(tagId)) {
+                GiftCertificateHasTag giftCertificateHasTag = GiftCertificateHasTag.builder()
+                        .giftCertificateId(giftCertificateId)
+                        .tagId(tagId)
+                        .build();
+                giftCertificateHasTagDao.create(giftCertificateHasTag);
+            }
+        }
+    }
+
+    private Map<String, Object> findUpdateInfo(GiftCertificate giftCertificate) {
+        Map<String, Object> updateInfo = new HashMap<>();
+        carefullyValidateGiftCertificateValues(giftCertificate);
+        updateInfo.put(ColumnName.NAME, giftCertificate.getName());
+        updateInfo.put(ColumnName.DESCRIPTION, giftCertificate.getDescription());
+        updateInfo.put(ColumnName.PRICE, giftCertificate.getPrice());
+        updateInfo.put(ColumnName.DURATION, giftCertificate.getDuration());
+        return updateInfo;
+    }
+
+    @Override
+    @Transactional
+    public void deleteById(long id) {
+        Optional<GiftCertificate> certificateOptional = giftCertificateDao.findById(id);
+        if (!certificateOptional.isPresent()) {
+            throw new NotFoundEntityException("certificate.not.found");
+        }
+        giftCertificateDao.deleteById(id);
+    }
+
+    @Override
+    @Transactional
+    public List<GiftCertificateDto> findAllWithTags(String tagName, String partInfo,
+                                                    List<String> sortColumns, List<String> orderTypes) {
+        List<GiftCertificateDto> giftCertificateDtoList = new ArrayList<>();
+        List<GiftCertificate> giftCertificates;
+        if (sortColumns != null) {
+            SortParamsContext sortParamsContext = new SortParamsContext(sortColumns, orderTypes);
+            validateSortParamsContext(sortParamsContext);
+            if (tagName != null || partInfo != null) {
+                giftCertificates = findGiftCertificatesWithSortingAndFiltering(tagName,
+                        partInfo, sortParamsContext);
+            } else {
+                giftCertificates = giftCertificateDao.findAllWithSorting(sortParamsContext);
+            }
+        } else if (tagName != null || partInfo != null) {
+            giftCertificates = findCertificatesWithFiltering(tagName, partInfo);
+        } else {
+            giftCertificates = giftCertificateDao.findAll();
```

Comment:
1. is it a correct solution that if id and name are not null but empty or spaces?
2. is it correct solution that if id and name are not blank and we are going to return all possible values?